### DISCUSSION TITLE
Failed to build ruby 2.7.5 on ubuntu 22.04

### DISCUSSION THREAD

**Discussion by wheatley** [rbenv/ruby-build#1940] [Build failures] [answered]
When trying to install ruby 2.7.5 using `rbenv install 2.7.5`

OS: Ubuntu 22.04

```
Generating RI format into /tmp/ruby-build.20220222121734.74310.fnuTKg/ruby-2.7.5/.ext/rdoc...


  Files:       1144

  Classes:     1618 ( 759 undocumented)
  Modules:      328 ( 135 undocumented)
  Constants:   2435 ( 726 undocumented)
  Attributes:  1490 ( 559 undocumented)
  Methods:    13068 (4014 undocumented)

  Total:      18939 (6193 undocumented)
   67.30% documented

  Elapsed: 19.8s

generating x86_64-linux-fake.rb
x86_64-linux-fake.rb updated
./miniruby -I./lib -I. -I.ext/common  ./tool/runruby.rb --extout=.ext  -- --disable-gems -r./x86_64-linux-fake ./tool/rbinstall.rb --make="make" --dest-dir="" --extout=".ext" --mflags="" --make-flags="" --data-mode=0644 --prog-mode=0755 --installed-list .installed.list --mantype="doc" --rdoc-output=".ext/rdoc" --html-o
utput=".ext/html"
installing binary commands:         /home/vagrant/.rbenv/versions/2.7.5/bin
installing base libraries:          /home/vagrant/.rbenv/versions/2.7.5/lib
installing arch files:              /home/vagrant/.rbenv/versions/2.7.5/lib/ruby/2.7.0/x86_64-linux
installing pkgconfig data:          /home/vagrant/.rbenv/versions/2.7.5/lib/pkgconfig
installing extension objects:       /home/vagrant/.rbenv/versions/2.7.5/lib/ruby/2.7.0/x86_64-linux
installing extension objects:       /home/vagrant/.rbenv/versions/2.7.5/lib/ruby/site_ruby/2.7.0/x86_64-linux
installing extension objects:       /home/vagrant/.rbenv/versions/2.7.5/lib/ruby/vendor_ruby/2.7.0/x86_64-linux
installing extension headers:       /home/vagrant/.rbenv/versions/2.7.5/include/ruby-2.7.0/x86_64-linux
installing extension scripts:       /home/vagrant/.rbenv/versions/2.7.5/lib/ruby/2.7.0
installing extension scripts:       /home/vagrant/.rbenv/versions/2.7.5/lib/ruby/site_ruby/2.7.0
installing extension scripts:       /home/vagrant/.rbenv/versions/2.7.5/lib/ruby/vendor_ruby/2.7.0
installing extension headers:       /home/vagrant/.rbenv/versions/2.7.5/include/ruby-2.7.0/ruby
installing rdoc:                    /home/vagrant/.rbenv/versions/2.7.5/share/ri/2.7.0/system
installing html-docs:               /home/vagrant/.rbenv/versions/2.7.5/share/doc/ruby
installing capi-docs:               /home/vagrant/.rbenv/versions/2.7.5/share/doc/ruby
installing command scripts:         /home/vagrant/.rbenv/versions/2.7.5/bin
installing library scripts:         /home/vagrant/.rbenv/versions/2.7.5/lib/ruby/2.7.0
installing common headers:          /home/vagrant/.rbenv/versions/2.7.5/include/ruby-2.7.0
installing manpages:                /home/vagrant/.rbenv/versions/2.7.5/share/man (man1, man5)
installing default gems from lib:   /home/vagrant/.rbenv/versions/2.7.5/lib/ruby/gems/2.7.0 (build_info, cache, doc, extensions, gems, specifications)
                                    benchmark 0.1.0
/tmp/ruby-build.20220222121734.74310.fnuTKg/ruby-2.7.5/lib/rubygems/core_ext/kernel_require.rb:83:in `require': cannot load such file -- openssl (LoadError)
        from /tmp/ruby-build.20220222121734.74310.fnuTKg/ruby-2.7.5/lib/rubygems/core_ext/kernel_require.rb:83:in `require'
        from /tmp/ruby-build.20220222121734.74310.fnuTKg/ruby-2.7.5/lib/rubygems/specification.rb:2430:in `to_ruby'
        from ./tool/rbinstall.rb:846:in `block (2 levels) in install_default_gem'
        from ./tool/rbinstall.rb:279:in `open_for_install'
        from ./tool/rbinstall.rb:845:in `block in install_default_gem'
        from ./tool/rbinstall.rb:835:in `each'
        from ./tool/rbinstall.rb:835:in `install_default_gem'
        from ./tool/rbinstall.rb:799:in `block in <main>'
        from ./tool/rbinstall.rb:950:in `block in <main>'
        from ./tool/rbinstall.rb:947:in `each'
        from ./tool/rbinstall.rb:947:in `<main>'
make: *** [uncommon.mk:373: do-install-all] Error 1
```

**Comment by mislav** [comment]
Be sure to follow our [recommended build setup for your OS](https://github.com/rbenv/ruby-build/wiki#suggested-build-environment). For Ubuntu, I would ensure that you've installed at least `libssl-dev` before compiling Ruby.

**Reply by wheatley** [reply]
hey @mislav thanks for the reply,

yep tried installing libssl-dev

https://packages.ubuntu.com/search?keywords=libssl-dev

- [impish-updates](https://packages.ubuntu.com/impish-updates/libssl-dev) (libdevel): Secure Sockets Layer toolkit - development files
- 1.1.1l-1ubuntu1.1: amd64 arm64 armhf i386 ppc64el s390x
- [jammy](https://packages.ubuntu.com/jammy/libssl-dev) (libdevel): Secure Sockets Layer toolkit - development files
- 3.0.1-0ubuntu1: amd64 arm64 armhf i386 ppc64el s390x

Do you know of anyway to have libssl-dev1.1.1 as well as 3.01 installed together?

**Reply by mislav** [reply]
I think `1.1.1l` would be better. But sorry, I'm not too familiar with how Ubuntu packages things.

**Reply by eregon** [reply]
https://github.com/rbenv/ruby-build/discussions/1940#discussioncomment-2519546 so 1.1.x is not available as a system package on Ubuntu 22.04

**Comment by wheatley** [comment]
Hi, was wondering if there was an update to this error?

**Reply by stuffa** [reply]
This issue applies to all versions of ruby.  I have tried 2.7.5, 3.0.3 and 3.1.1

**Reply by hsbt** [reply]
Really? I could build Ruby 3.1 on Ubuntu 22.04.

**Reply by caseyrichins** [reply]
I was able to install ruby 3.2.1 AFTER removing GEM_HOME and disabling PATH="$HOME/gems/bin:$PATH" from my .bashrc file. Apparently when trying to install ruby it was looking for yaml in my gems directory, there was a yaml there but it was for gem version. This cause me no end of headache to figure after upgrading form 20.04 to 22.04. I ended up removing all ruby and gem and starting from a scratch rbenv install. I was only successfull after again disabling the old GEM_HOME stuff in $HOME/.bashrc

**Comment by hsbt** [comment, marked as answer]
Unfortunately, Ubuntu 22.04 only provides OpenSSL 3.0, which is supported only by Ruby versions 3.1+.

Older versions of Ruby need OpenSSL 1.1, which doesn't come with newer Ubuntu out of the box.

**Reply by mislav** [reply]
@wheatley @stuffa Based on this information, I would think that the safest way to build Ruby on Ubuntu 22.04 would be: instead of depending on the system `libssl-dev` package, download and compile OpenSSL 1.1.1 into its own directory, then point to that directory with `RUBY_CONFIGURE_OPTS=--with-openssl-dir=/path/to/my/openssl` while compiling with ruby-build.

**Reply by mislav** [reply]
Update: [newer versions of ruby-build](https://github.com/rbenv/ruby-build/releases/tag/v20220710) work around the problem of the user having to manually download and install the required version of OpenSSL.

What this means when installing, for example, Ruby 2.7.5 on Ubuntu 22.04 is that ruby-build will automatically download and compile OpenSSL 1.1 and link Ruby to that instead of to the system OpenSSL.

Make sure to upgrade your ruby-build!

**Reply by wheatley** [reply]
Thank you for the update @mislav and everyone that has helped each other with this issue.

**Comment by hsbt** [comment]
FYI: https://bugs.ruby-lang.org/issues/18658#note-4

**Comment by stuffa** [comment]
I found that I could compile 3.1.1 ok
So I migrated all my ruby projects to 3.1.1 and Rails 7 (I was going to do
it anyway, just was not planning to do it so soon)

Cheers

--------------------------------------------------------
Chris Martin
m: +61 419 812 371
e: ***@***.***
--------------------------------------------------------

**Comment by sudoyang** [comment]
I met exactly the same problem. Ruby 2.7 is essential for me to use for jekyll and github page since Ruby 3.x is incompatible with github page. I try to install Ruby 2.7 via rbenv in Ubuntu 22.04 and got this issue. 😂

**Reply by Sagor2214** [reply]
[]()

**Comment by romulogcerqueira** [comment]
I have faced the same issue.

**Reply by Sagor2214** [reply]
I have faced the same issue

**Comment by mchilson** [comment]
Facing the same issue here since upgrading to 22.04 LTS when trying to install 2.7.5 via rbenv.

Downloading ruby-2.7.5.tar.bz2...
-> https://cache.ruby-lang.org/pub/ruby/2.7/ruby-2.7.5.tar.bz2
Installing ruby-2.7.5...

BUILD FAILED (Ubuntu 22.04 using ruby-build 20211227-3-gcdc215e)

Inspect or clean up the working tree at /tmp/ruby-build.20220428175448.2543631.uzBfU1
Results logged to /tmp/ruby-build.20220428175448.2543631.log

Last 10 log lines:
	from ./tool/rbinstall.rb:846:in `block (2 levels) in install_default_gem'
	from ./tool/rbinstall.rb:279:in `open_for_install'
	from ./tool/rbinstall.rb:845:in `block in install_default_gem'
	from ./tool/rbinstall.rb:835:in `each'
	from ./tool/rbinstall.rb:835:in `install_default_gem'
	from ./tool/rbinstall.rb:799:in `block in <main>'
	from ./tool/rbinstall.rb:950:in `block in <main>'
	from ./tool/rbinstall.rb:947:in `each'
	from ./tool/rbinstall.rb:947:in `<main>'
make: *** [uncommon.mk:373: do-install-all] Error 1

**Comment by nielsonrolim** [comment]
Following the suggestion of @mislav, I compiled [OpenSSL 1.1.1](https://www.openssl.org/source/) and it worked. I tested on **Ubuntu 22.04** and **Fedora 36**.

**Install the dependencies**

**Ubuntu 22.04**:
```shell
$ sudo apt install build-essential checkinstall zlib1g-dev
```
**Fedora 36**:
```shell
$ sudo dnf groupinstall "Development Tools"
$ sudo dnf install perl-core zlib-devel
```
---
**Download OpenSSL 1.1.1**
```
$ cd ~/Downloads
$ wget https://www.openssl.org/source/openssl-1.1.1q.tar.gz
$ tar xf openssl-1.1.1q.tar.gz
```
---
**Compile it**
```
$ cd ~/Downloads/openssl-1.1.1q
$ ./config --prefix=/opt/openssl-1.1.1q --openssldir=/opt/openssl-1.1.1q shared zlib
$ make
$ make test
$ sudo make install
```
---
**Link the system's certs to OpenSSL 1.1.1 directory**
```
$ sudo rm -rf /opt/openssl-1.1.1q/certs
$ sudo ln -s /etc/ssl/certs /opt/openssl-1.1.1q
```
---
**Install ruby**

Use `RUBY_CONFIGURE_OPTS=--with-openssl-dir=/opt/openssl-1.1.1q` before the command to install the ruby version
```
$ RUBY_CONFIGURE_OPTS=--with-openssl-dir=/opt/openssl-1.1.1q rbenv install 2.7.6
```
It works with `asdf` as well:
```
$ RUBY_CONFIGURE_OPTS=--with-openssl-dir=/opt/openssl-1.1.1q asdf install ruby 2.7.6
```
If you want to make this permanent, add this line to your `.bashrc` or `.zshrc` file:
```
export RUBY_CONFIGURE_OPTS="--with-openssl-dir=/opt/openssl-1.1.1q/"
```
Then you don't need to use `RUBY_CONFIGURE_OPTS=--with-openssl-dir=/opt/openssl-1.1.1q` before the command anymore.

**Reply by thiagobarbosa-dev** [reply]
Man...worked for me with asdf too in Pop!_OS 22.04
Thank you so much

I have

**Reply by sudoyang** [reply]
I still got errors...
I installed openssl in a different location. Does this matter?
I can build openssl  and "libssl-dev libreadline-dev zlib1g-dev libyaml-dev" have been installed. But the install process seems to fail detecting them.

Downloading ruby-2.7.6.tar.bz2...
-> https://cache.ruby-lang.org/pub/ruby/2.7/ruby-2.7.6.tar.bz2
Installing ruby-2.7.6...

BUILD FAILED (Ubuntu 22.04 using ruby-build 20220415)

Inspect or clean up the working tree at /tmp/ruby-build.20220502133038.951139.t4CK79
Results logged to /tmp/ruby-build.20220502133038.951139.log

Last 10 log lines:
The Ruby yaml extension was not compiled.
ERROR: Ruby install aborted due to missing extensions
Try running `apt-get install -y libssl-dev libreadline-dev zlib1g-dev libyaml-dev` to fetch missing dependencies.

Configure options used:
  --prefix=/home/phy/.rbenv/versions/2.7.6
  --enable-shared
  --with-openssl-dir=/home/phy/Tools/openssl-1.1.1n/
  LDFLAGS=-L/home/phy/.rbenv/versions/2.7.6/lib 
  CPPFLAGS=-I/home/phy/.rbenv/versions/2.7.6/include

**Reply by nielsonrolim** [reply]
It should work as long as you add the path you installed to `RUBY_CONFIGURE_OPTS=--with-openssl-dir=`

**Reply by josydeleon** [reply]
Same issue here after following the steps by @nielsonrolim

**Reply by josydeleon** [reply]
Why homebrew? `mkdir: cannot create directory ‘/opt/homebrew’: Permission denied`

@sudoyang are you having the same message?

```
Downloading ruby-2.6.6.tar.bz2...
-> https://cache.ruby-lang.org/pub/ruby/2.6/ruby-2.6.6.tar.bz2
Installing ruby-2.6.6...
mkdir: cannot create directory ‘/opt/homebrew’: Permission denied

BUILD FAILED (Ubuntu 22.04 using ruby-build 20200401)

Inspect or clean up the working tree at /tmp/ruby-build.20220502200224.175907.RvDvm0
Results logged to /tmp/ruby-build.20220502200224.175907.log

Last 10 log lines:
/tmp/ruby-build.20220502200224.175907.RvDvm0 ~
/tmp/ruby-build.20220502200224.175907.RvDvm0/ruby-2.6.6 /tmp/ruby-build.20220502200224.175907.RvDvm0 ~
```

**Reply by sudoyang** [reply]
@jossydeleon My issue is different. It seems that your brew does not have permission to write /opt/. Maybe you can install openssl somewhere you can write. 

UPDATE:
Worked. I reinstall rbenv manually. There are system installation conflicts and the manual install solves the issue.

**Reply by santiagofavrin** [reply]
Thank you <3

**Reply by Samuelodan** [reply]
This worked perfectly. Thank you so much.

**Reply by mlchai** [reply]
Thanks so much for putting this out here, did the trick after searching around for a fix.

**Reply by cjamison** [reply]
Thanks @nielsonrolim! Although, I think there is a typo in the "Link the system's certs to OpenSSL 1.1.1 directory" section. Shouldn't the `ln` line be the following instead?
```
$ sudo ln -s /etc/ssl/certs /opt/openssl-1.1.1o/certs
```

**Reply by nielsonrolim** [reply]
It works either way. If you don't define the link destination name, it will use the same name of the origin directory or file. In that case it will be `certs`.

**Reply by cjamison** [reply]
Ahh, my apologies, this did not work for me initially so I must have mis-copied or mis-typed something. I thought the issue might have been the `ln` command but it was probably just human error on my part. Thanks!

**Reply by veragarciaenr** [reply]
@nielsonrolim This is great, thanks a lot!

**Reply by echeverriadev** [reply]
thank u bro! @nielsonrolim

**Reply by emmoth** [reply]
Worked for med, thanks :)

**Reply by allefgomes** [reply]
It's  a better solution.
Thankyou!!!
Worked for me in Ubuntu 22.04

**Reply by gabrielsaviank** [reply]
Worked like a charm, also this is the correct command in asdf 
`RUBY_CONFIGURE_OPTS=--with-openssl-dir=/opt/openssl-1.1.1o asdf install ruby 2.7.6`

**Reply by nielsonrolim** [reply]
> Worked like a charm, also this is the correct command in asdf `RUBY_CONFIGURE_OPTS=--with-openssl-dir=/opt/openssl-1.1.1o asdf install ruby 2.7.6`

Thanks! I updated my post

**Reply by gpalazuelosg** [reply]
Hi!, I confirm the suggested steps to compile openssl1.1.1 worked for me on a system with Ubuntu 22.04 configured today. Thank you!

**Reply by foxmike** [reply]
Thanks nielsonrolim!  Worked for me 2.7.5

**Reply by plaindocs** [reply]
I'm getting this error on Fedora

```
ripper.c: In function ‘ripper_yyparse’:
ripper.c:6309:12: error: ‘YYEMPTY’ undeclared (first use in this function)
 6309 |   yychar = YYEMPTY; /* Cause a token to be read.  */
      |            ^~~~~~~
ripper.c:6309:12: note: each undeclared identifier is reported only once for each function it appears in
ripper.c:6441:22: error: ‘YYerror’ undeclared (first use in this function); did you mean ‘yyerror’?
 6441 |     {
      |                      ^      
      |                      yyerror
ripper.c:6447:16: error: ‘YYUNDEF’ undeclared (first use in this function); did you mean ‘T_UNDEF’?
 6447 |     {
      |                ^      
      |                T_UNDEF
make[2]: *** [Makefile:289: ripper.o] Error 1
```

**Reply by GusChenn** [reply]
Worked for me, thansk a lot!

**Reply by theadzik** [reply]
Worked great, thanks. Do I need to undo this part after the installation to use 3.0 again?

```bash
$ sudo rm -rf /opt/openssl-1.1.1q/certs
$ sudo ln -s /etc/ssl/certs /opt/openssl-1.1.1q
```

**Reply by nielsonrolim** [reply]
No, no need. You can keep both system's openssl and `/opt/openssl-1.1.1q`.
If you want to install ruby 3.0 using sysstem's openssl, just do not use `RUBY_CONFIGURE_OPTS=--with-openssl-dir=/opt/openssl-1.1.1q` before the install command.

**Reply by shanesmith** [reply]
@plaindocs I'm getting the same `YYEMPTY` error, were you able to fix it on your side?

**Reply by plaindocs** [reply]
It's been a while, but I believe this https://github.com/rbenv/ruby-build/pull/1974 fixed things for me?

**Reply by nielsonrolim** [reply]
Yeah, it fixed for me too

**Reply by rreddy693** [reply]
If in case anyone trying to install ruby with rvm then follow the above mentioned steps but at end while installing ruby follow below command in ubuntu 22.04

rvm install ruby-2.5.1 --with-openssl-dir=/opt/openssl-1.1.1q

**Reply by Nicolas-Hermet** [reply]
Thank you so much for this @rreddy693 : worked for me !!!

**Comment by max-las** [comment]
I stumbled upon this issue today as well. 
I had installed rbenv with homebrew (on ubuntu 22.04), and it turns out openssl 1.1.1n was already present through homebrew after that, as I could see by running `openssl version` and `which openssl`. So I just ran :
```
RUBY_CONFIGURE_OPTS=--with-openssl-dir=/home/linuxbrew/.linuxbrew/opt/openssl@1.1 rbenv install 2.7.6
```
And it worked.

**Reply by antonzaytsev** [reply]
worked for me, thank you!

**Reply by SahSantoshh** [reply]
Thanks this worked 👍🏼

**Comment by mchilson** [comment]
Thanks... that solution worked perfect.

**Comment by IvanDerlich** [comment]
Worked for me!

**Comment by sas2job** [comment]
Does not work for 2.7.3 but work for 3.1.2
```console
RUBY_CONFIGURE_OPTS=--with-openssl-dir=/opt/openssl-1.1.1o rbenv install 2.7.3
BUILD FAILED (Ubuntu 22.04 using ruby-build 20220426-3-g1038c07)

Inspect or clean up the working tree at /tmp/ruby-build.20220522135821.31657.Iqe8Ck
Results logged to /tmp/ruby-build.20220522135821.31657.log

Last 10 log lines:
linking shared-object bigdecimal.so
make[2]: Leaving directory '/tmp/ruby-build.20220522135821.31657.Iqe8Ck/ruby-2.7.3/ext/bigdecimal'
ripper.y: At top level:
cc1: note: unrecognized command-line option ‘-Wno-self-assign’ may have been intended to silence earlier diagnostics
cc1: note: unrecognized command-line option ‘-Wno-parentheses-equality’ may have been intended to silence earlier diagnostics
cc1: note: unrecognized command-line option ‘-Wno-constant-logical-operand’ may have been intended to silence earlier diagnostics
linking shared-object ripper.so
make[2]: Leaving directory '/tmp/ruby-build.20220522135821.31657.Iqe8Ck/ruby-2.7.3/ext/ripper'
make[1]: Leaving directory '/tmp/ruby-build.20220522135821.31657.Iqe8Ck/ruby-2.7.3'
make: *** [uncommon.mk:295: build-ext] Error 2
```

**Reply by xjunior** [reply]
@sas2job did you solve this `uncommon.mk:295` error?

**Reply by sas2job** [reply]
Using vbox & ubuntu 20.04

**Comment by t-yanick** [comment]
Worked for me.
Thanks very much

**Comment by tomjuggler** [comment]
rbenv install 2.7.5 gives me "ruby-build: definition not found: 2.7.5"

not able to find 2.7.5 with "rbenv install --list"

Also need this for Github Pages..

**Comment by brianpursley** [comment]
I got the same error trying to use rbenv to install 2.7.1, and [part of this Ask Ubuntu answer](https://askubuntu.com/a/1407024/866464) helped me.  I'll leave this here in case it helps someone else.

Here is what I did:
```
wget http://security.ubuntu.com/ubuntu/pool/main/o/openssl/openssl_1.1.1f-1ubuntu2.16_amd64.deb
sudo dpkg -i openssl_1.1.1f-1ubuntu2.16_amd64.deb

wget http://security.ubuntu.com/ubuntu/pool/main/o/openssl/libssl-dev_1.1.1f-1ubuntu2.16_amd64.deb
sudo dpkg -i libssl-dev_1.1.1f-1ubuntu2.16_amd64.deb
```

I'm not sure if both of those are needed, or if only libssl-dev is needed, but after installing both of those, I was able to successfully install Ruby 2.7.1 using `rbenv install 2.7.1`

---

Also, I should mention, the reason I need Ruby 2.7.1 is because [Github Pages does not support Jekyll 4](https://github.com/github/pages-gem/issues/651) and Jekyll 3 requires Ruby 2.x.  I suspect others are in the same situation.

---

Edit to add:

The next time I did `sudo apt upgrade` libssl-dev and openssl both got upgraded.  I think it's fine (actually preferable) as I don't need them anymore once Ruby 2.7.1 is installed/compiled, I think... :sweat_smile:

**Reply by nielsonrolim** [reply]
Today I installed ruby 2.7.6 using `asdf` without compiling openssl manually.

`asdf` downloaded openssl 1.1.1 and installed in its own directory.

I don't know if this is a correction implemented by `asdf` or `rbenv`.

**Comment by rreddy693** [comment]
If in case anyone trying to install ruby with rvm then follow the above mentioned steps but at end while installing ruby follow below command in ubuntu 22.04

`rvm install ruby-2.5.1 --with-openssl-dir=/opt/openssl-1.1.1q`
